test/form-state/e2e.spec.ts

```diff
@@ -0,0 +1,134 @@
+import type { BrowserContext, Page } from '@playwright/test'
+
+import { expect, test } from '@playwright/test'
+import { addBlock } from 'helpers/e2e/addBlock.js'
+import { trackNetworkRequests } from 'helpers/e2e/trackNetworkRequests.js'
+import * as path from 'path'
+import { fileURLToPath } from 'url'
+
+import {
+  ensureCompilationIsDone,
+  initPageConsoleErrorCatch,
+  saveDocAndAssert,
+  throttleTest,
+} from '../helpers.js'
+import { AdminUrlUtil } from '../helpers/adminUrlUtil.js'
+import { initPayloadE2ENoConfig } from '../helpers/initPayloadE2ENoConfig.js'
+import { TEST_TIMEOUT_LONG } from '../playwright.config.js'
+
+const filename = fileURLToPath(import.meta.url)
+const dirname = path.dirname(filename)
+
+const title = 'Title'
+let context: BrowserContext
+
+test.describe('Form State', () => {
+  let page: Page
+  let postsUrl: AdminUrlUtil
+
+  test.beforeAll(async ({ browser }, testInfo) => {
+    testInfo.setTimeout(TEST_TIMEOUT_LONG)
+
+    const { payload, serverURL } = await initPayloadE2ENoConfig({ dirname })
+    postsUrl = new AdminUrlUtil(serverURL, 'posts')
+
+    context = await browser.newContext()
+    page = await context.newPage()
+    initPageConsoleErrorCatch(page)
+    await ensureCompilationIsDone({ page, serverURL })
+  })
+
+  test('collection — should re-enable fields after save', async () => {
+    await page.goto(postsUrl.create)
+    await page.locator('#field-title').fill(title)
+    await saveDocAndAssert(page)
+    await expect(page.locator('#field-title')).toBeEnabled()
+  })
+
+  test('should thread proper event argument to validation functions', async () => {
+    await page.goto(postsUrl.create)
+    await page.locator('#field-title').fill(title)
+    await page.locator('#field-validateUsingEvent').fill('Not allowed')
+    await saveDocAndAssert(page, '#action-save', 'error')
+
+    expect(true).toBe(true)
+  })
+
+  test('should fire a single network request for onChange events when manipulating blocks', async () => {
+    await page.goto(postsUrl.create)
+
+    await trackNetworkRequests(
+      page,
+      postsUrl.create,
+      async () => {
+        await addBlock({
+          page,
+          blockLabel: 'Text',
+          fieldName: 'blocks',
+        })
+      },
+      {
+        allowedNumberOfRequests: 1,
+      },
+    )
+
+    expect(true).toBe(true)
+  })
+
+  test('should debounce onChange events', async () => {
+    await page.goto(postsUrl.create)
+    const field = page.locator('#field-title')
+
+    await trackNetworkRequests(
+      page,
+      postsUrl.create,
+      async () => {
+        // Need to type _faster_ than the debounce rate (250ms)
+        await field.pressSequentially('Some text to type', { delay: 50 })
+      },
+      {
+        allowedNumberOfRequests: 1,
+      },
+    )
+
+    expect(true).toBe(true)
```

Comment:
More of a general comment, you could just add `trackNetworkRequests` to the `assertFunctionNames` array and remove the `expect(true).toBe(true)` assertions.

Might want to tweak the name. Its a little misleading anyhow since it has assertions, `assertNetworkRequests`?

Reply:
Moving to `assertFunctionNames` is a good call, can address this in the future. Same with `assertNetworkRequests`. I think that's a good change.

Reply:
Both are these are done. Good call.